Bugfix - Remove Non Rrule Properties

marcrd opened the pull request:
### Overview

This fix removes the error that happens when editing floating time. Unfortunately, the way the library is set up, makes it very hard to decorate the library. Instead forked the latest version with proper fixes. 

The way this works is we take the set of main recurrence properties by string, and remove anything that is not an `RRULE`.


### Changelog

- Creates a private function that matches for `RRUL`E and removes any other recurrence properties
- Fixes typo where it says `unsupported RDATE parm` for `EXDATE`, as it was confusing when debugging


![recurrence_irl](https://media.giphy.com/media/xT5LMHjKqPYWeeVpAc/giphy.gif)

marcrd committed:
Parses out floating time zone values

marcrd commented:
Tagged in @brianmuse for tracking

zachdunn commented on lib/rrule.js:
```diff
@@ -2209,13 +2209,39 @@
+        var index = parsedRecurrenceString.indexOf(ruleString);
+
+        if (contains(ruleString, 'TZID')) {
+          if (index !== -1) {
```
You can consolidate this and the encapsulating `if`

zachdunn commented on lib/rrule.js:
```diff
@@ -2209,13 +2209,39 @@
+
+      // Remove any of the recurrence date properties that contain floating timezone properties.
+      parsedRecurrenceString.forEach(function (ruleString) {
+        var index = parsedRecurrenceString.indexOf(ruleString);
```
`rruleIndex` would be more readable.

zachdunn commented on lib/rrule.js:
```diff
@@ -2209,13 +2209,39 @@
 
+    /**
+     * Filters out floating timezone properties.
+     * @param {String} s A string containing the set of recurrence properties.
```
We don't have to use the shorthand param convention in this fork.

zachdunn commented on lib/rrule.js:
```diff
@@ -2209,13 +2209,39 @@
       var defaultKeys = Object.keys(RRuleStr.DEFAULT_OPTIONS)
 
+      // Sets the recurrence string to the original recurrence model without any floating timezones.
+      s = this.filterFloatingTimezones(s)
```
Rather than mutating the original param, why don't we keep it handy? Think:

```javascript
parse: function (rawString, options) {
// ...
// Terrible name, but you get idea.
var safeString = this.filterFloatingTimezones(rawString);
```

marcrd committed:
Cleanup and naming convention updates
- Combines if statements insides filter function
- Updates naming conventions to not used shorthand
- Instead of mutating recurrence string, sets new variable
to prevent mutation and keeps original param intact

marcrd committed:
Updates param name`

marcrd committed:
Updates docblock to reflect iCal spec

marcrd committed:
Make sure to set var

jutaz commented on lib/rrule.js:
```diff
@@ -2209,13 +2209,40 @@
+     */
+    filterFloatingTimezones: function (stringRule) {
+      // Uses a regex to split the recurrence string by carriage return.
+      var parsedRecurrenceString = stringRule.split(/\s+/)
```
Are we OK with this also matching any spaces? because it will match any spaces too, not just newlines

marcrd replied:
Yes, as far as I can tell newlines should be the only separator of these params. You shouldn't have something the recurrence params being separated by a space in this library. If you have a regex that only splits by carriage return Im all ears though

jutaz replied:
It would be as simple regex as `/[\n\r]/`

marcrd committed:
Split by newline regex

marcrd commented:
@jutaz updated

jutaz reviewed (changes requested):
lets also add tests to this

marcrd committed:
Remove semi colon

marcrd commented:
After conversation with @jutaz 

> Hmmm this library doesn't really do "functional" tests for their private functions
> The tooling isn't really built for me to easily assert against my new function. I could but the libraries current testing doesn't really establish any precedent for it and would probably need updating to better assert against functional tests as it only uses mocha

Henry-L commented:
should we bump the version up too?

marcrd committed:
Bump versioning

brianmuse commented:
Doesn't this actually introduce the feature as a bug into the library? Stripping out the TZID just leaves us with a datetime without any context of when it actually takes place. If the goal is to avoid the library crashing, than I'm not sure this is the right place to do this.

marcrd committed:
Makes copy on for each and splices anything but rrule

marcrd commented on lib/rrule.js:
```diff
@@ -2209,13 +2209,41 @@
+      var parsedRecurrenceString = stringRule.split(newline)
+
+      // Remove any of the recurrence date properties that contain floating timezone properties.
+      parsedRecurrenceString.slice(0).forEach(function (ruleString) {
```
The slice here is to make sure we do this on a copy so we don't skip over values

atticoos replied:
Seems a lot simpler to just filter

```js
return parsedRecurrenceString.filter(ruleString => contains(ruleString, 'RRULE')).join('\n')
```

marcrd replied:
Sorry meant to close this

marcrd commented:
Closing this in favor of https://github.com/robinpowered/robin-dashboard/pull/4817/commits/a49f73123b032d4ecf6eefd6b6a862caebcd51ce

Per @brianmuse we're going to create our own submodels with parsed recurrence rules rather than filter it out here

atticoos commented on lib/rrule.js:
```diff
@@ -2209,13 +2209,41 @@
+      parsedRecurrenceString.slice(0).forEach(function (ruleString) {
+        var rruleIndex = parsedRecurrenceString.indexOf(ruleString)
+
+        if (!contains(ruleString, 'RRULE') && rruleIndex !== -1) {
```
General question -- so the lack of "RRULE" indicates its a floating timezone? From `3.3.5` it sounded like any date that lacks a timezone 

> DATE-TIME values of this type are said to be "floating" and are not bound to any time zone in particular.

atticoos replied:
Or is this actually just omitting entries like `RDATE` and `EXDATE` etc? If so, it might be worth doing that in the dependent, rather than the dependency, such that omit everything but the RRULE in the string _before_ we give it to the RRule library (versus doing that here)

marcrd replied:
@ajwhite well the idea was originally to parse out floating timezones, but now we changed it to parse anything besides `rrule` from recurrence. We're no longer fixing this in library and will move it to Robin

marcrd replied:
@ajwhite see https://github.com/robinpowered/robin-dashboard/pull/4817/commits/a49f73123b032d4ecf6eefd6b6a862caebcd51ce

atticoos replied:
cool, makes a lot more sense to apply that business logic in the project its required in 👍

marcrd replied:
Yeah it's just kinda not so great that you have to essentially modify the event submodels and create your own client side model for unsupported recurrence props, just because the library doesn't support on part of the spec
